File: solutions/C/sieve/4/sieve.c
#include "sieve.h"
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h> 

/**
 * @brief Finds prime numbers up to a given limit using the Sieve of Eratosthenes algorithm.
 *
 * @param limit The upper limit for finding prime numbers.
 * @param primes Pointer to an array where prime numbers will be stored.
 * @param max_primes The maximum number of primes to find.
 * @return The number of prime numbers found within the specified limit.
 */
uint32_t sieve(uint32_t limit, uint32_t *primes, size_t max_primes) {
    // Dynamically allocate memory for the is_prime array
    bool *is_prime = (bool *)malloc((limit + 1) * sizeof(bool));
    if (is_prime == NULL) {
        // Error handling for memory allocation failure
        return 0;
    }

    // Set all elements of is_prime array to true initially
    for (uint32_t i = 0; i <= limit; i++) {
        is_prime[i] = true;
    }

    uint32_t count = 0; // Counter for the number of prime numbers found
    for (uint32_t i = 2; i <= limit && count < max_primes; i++) {
        if (is_prime[i]) {
            // Optimized loop to mark multiples of prime numbers as non-prime
            for (uint32_t j = i * i; j <= limit; j += i) {
                is_prime[j] = false;
            }
            primes[count++] = i; // Store prime number and increment count
        }
    }

    free(is_prime); // Free dynamically allocated memory
    return count;
}

File: solutions/Python/sieve/1/sieve.py
from math import isqrt


def primes(limit: int) -> list[int]:
    if limit < 2:
        return []
    if limit == 2:
        return [2]
    is_prime = [True] * (limit + 1)
    is_prime[0] = False
    is_prime[1] = False

    for i in range(2, isqrt(limit) + 1):  
        if is_prime[i]:
            for x in range(i * i, limit + 1, i):
                is_prime[x] = False

    return [i for i in range(limit + 1) if is_prime[i]]
